src/Compilers/VisualBasic/Portable/Semantics/TypeInference/TypeArgumentInference.vb

```diff
@@ -1099,17 +1101,52 @@ HandleAsAGeneralExpression:
 
                         Dim possiblyGenericType = DirectCast(parameterType, NamedTypeSymbol)
 
-                        Do
-                            For Each typeArgument In possiblyGenericType.TypeArgumentsWithDefinitionUseSiteDiagnostics(Me.UseSiteDiagnostics)
-                                AddTypeToGraph(typeArgument, argNode, isOutgoingEdge, haveSeenTypeParameters)
+                        Dim elementTypes As ImmutableArray(Of TypeSymbol) = Nothing
+                        If possiblyGenericType.TryGetElementTypesIfTupleOrCompatible(elementTypes) Then
+                            For Each elementType In elementTypes
+                                AddTypeToGraph(elementType, argNode, isOutgoingEdge, haveSeenTypeParameters)
                             Next
+                        Else
+
+                            Do
+                                For Each typeArgument In possiblyGenericType.TypeArgumentsWithDefinitionUseSiteDiagnostics(Me.UseSiteDiagnostics)
+                                    AddTypeToGraph(typeArgument, argNode, isOutgoingEdge, haveSeenTypeParameters)
+                                Next
 
-                            possiblyGenericType = possiblyGenericType.ContainingType
-                        Loop While possiblyGenericType IsNot Nothing
+                                possiblyGenericType = possiblyGenericType.ContainingType
+                            Loop While possiblyGenericType IsNot Nothing
+                        End If
                 End Select
 
             End Sub
 
+            Private Sub AddTupleLiteralToGraph(argNode As ArgumentNode)
+                AddTupleLiteralToGraph(argNode.ParameterType, argNode)
+            End Sub
+
+            Private Sub AddTupleLiteralToGraph(
+                parameterType As TypeSymbol,
+                argNode As ArgumentNode
+            )
+                Debug.Assert(argNode.Expression.Kind = BoundKind.TupleLiteral)
+
+                Dim tupleLiteral = DirectCast(argNode.Expression, BoundTupleLiteral)
+
+                ' If literal has type, just infer from that type
+                If tupleLiteral.Type IsNot Nothing Then
+                    AddTypeToGraph(argNode, isOutgoingEdge:=True)
+                    Return
+                End If
```

Comment:
Vlad and I met in my office to discuss the kinds of unit tests needed to demonstrate my point. For example
M((T, U) tuple) with an argument (1, 2). The expressions 1 and 2 will not propagate to give types for T and U because the following loop is skipped. I just placed a brief comment to remind him where in the code we were discussing the issue.

---
In reply to: [73606403](https://github.com/dotnet/roslyn/pull/12932#discussion_r73606403) [](ancestors = 73606403)